tests/test_Proxy.py

```diff
@@ -0,0 +1,306 @@
+import pytest
+import asyncio
+from starkware.starknet.testing.starknet import Starknet
+from utils import (
+    Signer, assert_revert, assert_event_emitted, get_contract_def, cached_contract
+)
+
+# contract paths
+account_path = 'contracts/Account.cairo'
+proxy_path = 'contracts/proxy/Proxy.cairo'
+dummy_path = 'tests/mocks/dummy_implementation.cairo'
+dummy_path_v2 = 'tests/mocks/dummy_implementation_v2.cairo'
+
+# random value
+VALUE_1 = 123
+VALUE_2 = 987
+
+
+signer = Signer(123456789987654321)
+
+
+@pytest.fixture(scope='module')
+def event_loop():
+    return asyncio.new_event_loop()
+
+
+@pytest.fixture(scope='module')
+def contract_defs():
+    account_def = get_contract_def(account_path)
+    dummy_v1_def = get_contract_def(dummy_path)
+    dummy_v2_def = get_contract_def(dummy_path_v2)
+    proxy_def = get_contract_def(proxy_path)
+
+    return account_def, dummy_v1_def, dummy_v2_def, proxy_def
+
+
+@pytest.fixture(scope='module')
+async def proxy_init(contract_defs):
+    account_def, dummy_v1_def, dummy_v2_def, proxy_def = contract_defs
+    starknet = await Starknet.empty()
+    account1 = await starknet.deploy(
+        contract_def=account_def,
+        constructor_calldata=[signer.public_key]
+    )
+    account2 = await starknet.deploy(
+        contract_def=account_def,
+        constructor_calldata=[signer.public_key]
+    )
+    v1 = await starknet.deploy(
+        contract_def=dummy_v1_def,
+        constructor_calldata=[]
+    )
+    v2 = await starknet.deploy(
+        contract_def=dummy_v2_def,
+        constructor_calldata=[]
+    )
+    v3 = await starknet.deploy(
+        contract_def=dummy_v2_def,
+        constructor_calldata=[]
+    )
+    proxy = await starknet.deploy(
+        contract_def=proxy_def,
+        constructor_calldata=[v1.contract_address]
+    )
+    return (
+        starknet.state,
+        account1,
+        account2,
+        v1,
+        v2,
+        v3,
+        proxy
+    )
+
+
+@pytest.fixture
+def proxy_factory(contract_defs, proxy_init):
+    account_def, dummy_v1_def, dummy_v2_def, proxy_def = contract_defs
+    state, account1, account2, v1, v2, v3, proxy = proxy_init
+    _state = state.copy()
+    account1 = cached_contract(_state, account_def, account1)
+    account2 = cached_contract(_state, account_def, account2)
+    v1 = cached_contract(_state, dummy_v1_def, v1)
+    v2 = cached_contract(_state, dummy_v2_def, v2)
+    v3 = cached_contract(_state, dummy_v2_def, v3)
+    proxy = cached_contract(_state, proxy_def, proxy)
+
+    return account1, account2, v1, v2, v3, proxy
+
+
+@pytest.fixture
+async def after_upgrade(proxy_factory):
```

Comment:
I'm unsure about this. Would you agree in that the Proxy is just a proxy and the upgrades belong to the implementations? These could be generic proxies. If we follow this idea, maybe some of these tests belong to the upgrades library, which would also imply renaming the `proxy/` directory to `upgrades/` and have `upgrades/library.cairo` and `upgrades/Proxy.cairo` as components.

Also, it might make sense to have a specific implementation/upgradeable.cairo contract just to test initialization, upgrades, etc, and leave ERC20_Upgradeable just as an example preset. What do you think?

Reply:
This would be following the [UUPS Proxy standard](https://eips.ethereum.org/EIPS/eip-1822), which is a very specific upgrade proxy.

Reply:
> I'm unsure about this. Would you agree in that the Proxy is just a proxy and the upgrades belong to the implementations? These could be generic proxies. 

Great point! Yes, agreed on all fronts. 

> Also, it might make sense to have a specific implementation/upgradeable.cairo contract just to test initialization, upgrades, etc, and leave ERC20_Upgradeable just as an example preset. What do you think?

Looking back, I should have named `test_Proxy` to `test_upgrades` or something like that. But yes, I think your idea is a better approach.